sql/table_cache.h

```diff
@@ -236,7 +268,14 @@ class Table_cache_element {
       TABLE_list;
 
   TABLE_list used_tables;
-  TABLE_list free_tables;
+  /**
+    List of unused TABLE objects for tables without triggers or unused TABLE
+    objects for which triggers were not fully-loaded, so they can only can be
+    used by read-only statements.
+  */
+  TABLE_list free_tables_slim;
```

Comment:
The idea is to avoid scanning the whole (ok, half) of free_tables list to find a TABLE instance without triggers loaded when we are doing read-only statement. Splitting free_tables list in two - one with triggers and one without them allows to quickly check if unused TABLE with/without triggers exist and retrieve one for usage.